### PR TITLE
SentencePiece enhancements for Bergamot

### PR DESCRIPTION
**Pull request by jerinphilip**
The following is what's essentially getting added.

```cpp
void SentencePieceVocab::encodeAsSentencePieceText(const std::string& line,
                                                   sentencepiece::SentencePieceText& spt) const {
  spm_->Encode(line, &spt);
}
```

But I need compiler to know `SentencePieceVocab` for using `tryAs` for more than `FactoredVocab`:

```cpp
sentencepiece::SentencePieceText Tokenizer::tokenize(std::string const &snt) {
  // TODO(jerin): Bunch of hardcode here, 1, 0, need to get rid off somehow.
  auto vImpl = vocabs_[0]->tryAs<marian::SentencePieceVocab>();
  sentencepiece::SentencePieceText spt;
  vImpl->encodeAsSentencePieceText(snt, spt);
  return spt;
}
```

Therefore `sentencepiece_vocab.cpp` -> `sentencepiece_vocab.h` + `sentencepiece_vocab.cpp`.

### CODE REVIEW

File: src/data/sentencepiece_vocab.h
Comment (kpu) [inline src/data/sentencepiece_vocab.h:96]: Always have a trailing newline.

File: src/data/sentencepiece_vocab.h
Comment (kpu) [inline src/data/sentencepiece_vocab.h:86]: Why not just make a default implementation that throws/aborts/errors in the virtual parent?  Then you wouldn't need to move everything to the header.  And you're already doing a dynamic cast to get here, which is arguably worse.  
Even better would be processing into whatever byte aligned data structure you need, to include processing SentencePieceText within the method here and never exposing it.

File: src/data/sentencepiece_vocab.cpp
Comment (kpu) [inline src/data/sentencepiece_vocab.cpp:257]: There's a lot of unnecessary changes in this diff.  Please remove them.

**Review by kpu — CHANGES_REQUESTED** [review]
Make a virtual method in the parent?

File: src/data/sentencepiece_vocab.h
Comment (jerinphilip) [reply src/data/sentencepiece_vocab.h:86]: Honestly, I have been trying to avoid the dynamic cast (which is already in the code, probably for `FactoredVocab`), which is what nearly cost me a day. But this method is only SentencePieceVocab as it stands now. 

For what you're proposing, is the intention to future-proof normalizing nature of other possible SentencePiece alternatives?
In that case, I can abstract the parent to return a common implementation for the byte-aligned data structure (for marian vocabs) restricting sentencepiece dependency to just the SenteneceVocab, which I think is an additional advantage? 

In terms of naming `encodeAsSentencePieceText` -> `encodeWithStringRef`?

**Comment by ugermann** [comment]
1. I agree with Kenneth on the virtual method in the parent. We can ABORT for other vocab classes for now; implement them later if we want them. There should still be a compile guard checking for -DUSE_SENTENCEPIECE, because if the option is there, it might be used by someone (e.g. as part of regression testing, though I'm not sure) and then cause unnecessary back-and-forth about failing tests / compilation problems. 
2. Since C++ offers function overloading, I don't see the need for the long method name.
3. The method should take a sentencepiece::min_string_view (not referenced), not a std::string.

**Comment by jerinphilip** [comment]
> 1. I agree with Kenneth on the virtual method in the parent. We can ABORT for other vocab classes for now; implement them later if we want them. There should still be a compile guard checking for -DUSE_SENTENCEPIECE, because if the option is there, it might be used by someone (e.g. as part of regression testing, though I'm not sure) and then cause unnecessary back-and-forth about failing tests / compilation problems.
> 2. Since C++ offers function overloading, I don't see the need for the long method name.
> 3. The method should take a sentencepiece::min_string_view (not referenced), not a std::string.

**Comment by ugermann** [comment]
> > 1. The method should take a sentencepiece::min_string_view (not referenced), not a std::string.
> 
> Where is this coming from? future-ssplit?

It appears that the name has changed since I last checked. It's from sentencepiece_processor.h and now apparently called absl::string_view. In any case, we just need *some* form of string_view, and there are several out there, and it isn't part of C++ until C++-17 or so. I suggest we typedef it for mts as marian::StringView or marian::server::StringView and use that consistently.

I suggest that for the time being, we just use whatever comes with sentencepiece in sentencepiece_processor.h.

**Comment by jerinphilip** [comment]
> You are on the wrong track here. We do NOT want this function. The whole idea here is to avoid extra computation. What we want is
> 
> 1. Tokenization of sentence as a SentencePieceText before the sentence joins the queue (because we need length in tokens before queuing)
> 2. Conversion of that into a SentenceTuple at batch creation time. This is unfortunate but unavoidable unless we want to rewrite major parts of Marian.

Thanks for taking time out to review. 

Can you clarify? Once consumed as SourceToken (with just the information we need to map it back to the reference paragraph), there won't be any `SentencePiece*`. What @kpu suggested is to keep `SentencePiece*` hidden for further use in Marian, and known only to the `SentencePieceVocab` class. 

> Even better would be processing into whatever byte aligned data structure you need, to include processing SentencePieceText within the method here and never exposing it.

As I understand we cannot get byte-alignment without `spm_->Encode(...)` (is the other case possible?). This optimization might be too premature and can be moved around later to prevent any waste compute when we have the other components working. If you notice `string_view` is still pending until I get `SentencePiece*` to work. Am I missing something here?

Where is the extra computation? Is it a _lazy_-tokenization until the batch construction is in the works that is being implied here?

**Comment by kpu** [comment]
`SentenceTuple` expects `Words` so you should return `Words`, not a data structure we have to copy again to get `Words`.  

There is no such thing as `string_view` pending.  Do it now.  Because your function will take `string_view` in and return, as secondary output, a `std::vector<string_view> &alignments` such that alignments[i] is the original text for words[i].

**Comment by kpu** [comment]
> You are on the wrong track here. We do NOT want this function. 

We do want this function.  Just with an API that has `string_view` and produces `Words`.

File: src/data/sentencepiece_vocab.cpp
Comment (kpu) [inline src/data/sentencepiece_vocab.cpp:221]: We need Words to come out.  Don't roll it into the same vector.  Just give alignment as side information.

File: src/data/vocab.h
Comment (kpu) [inline src/data/vocab.h:54]: Let's get the `string_view` equivalent

**Comment by kpu** [comment]
The CI is breaking because the header path needs to include `marian-dev/src/3rd_party/sentencepiece/third_party/protobuf-lite`

You need a version of this that is scoped to just the file including it `sentencepiece.pb.h`
https://github.com/marian-nmt/sentencepiece/blob/8336bbd0c1cfba02a879afe625bf1ddaf7cd93c5/src/CMakeLists.txt#L50

**Comment by jerinphilip** [comment]
> The CI is breaking because the header path needs to include `marian-dev/src/3rd_party/sentencepiece/third_party/protobuf-lite`
> 
> You need a version of this that is scoped to just the file including `sentencepiece.pb.h`
> https://github.com/marian-nmt/sentencepiece/blob/8336bbd0c1cfba02a879afe625bf1ddaf7cd93c5/src/CMakeLists.txt#L50
> 
> And add it to `marian-dev/src/CMakeLists.txt`

I turned off `builtin_pb` altogether and is using generated protobuf at the moment, does this work for a solution. Ubuntu CI is passing. Mac seems to have protobuf issues. Windows is failing due to MKL.

**Comment by kpu** [comment]
Have you effectively added full protobuf as a dependency?

**Comment by jerinphilip** [comment]
> Have you effectively added full protobuf as a dependency?

Umm.. The following relevant line as long means as long as SentencePiece is in. Is this undesirable? 

https://github.com/browsermt/marian-dev/blob/751cdd72a17a73b41ac9e72950ba3a9043896063/src/3rd_party/CMakeLists.txt#L72

**Comment by jerinphilip** [comment]
So I'm working with `absl:::string_view` as `string_view` at the moment. Something works with the desired signature @kpu has provided, I'm about to tweak bergamot/mts with this. For this, **ssplit-cpp** currently works with [`pcrepp::StringPiece`](https://github.com/kpu/ssplit-cpp/blob/master/src/ssplit/ssplit.h#L19) at the moment. Conversion or just change **ssplit-cpp** with `string_view`?

File: src/data/sentencepiece_vocab.cpp
Comment (kpu) [inline src/data/sentencepiece_vocab.cpp:227]: Two lines.

File: src/3rd_party/CMakeLists.txt
Comment (kpu) [inline src/3rd_party/CMakeLists.txt:72]: This adds a dependency on system protobuf.

File: src/data/vocab.h
Comment (jerinphilip) [inline src/data/vocab.h:15]: Note to self, delete.

File: src/data/sentencepiece_vocab.cpp
Comment (kpu) [inline src/data/sentencepiece_vocab.cpp:231]: I think this should be piece.end() - piece.begin()

File: src/data/vocab.h
Comment (kpu) [inline src/data/vocab.h:9]: Why not put it in namespace marian?

File: src/data/vocab_base.h
Comment (kpu) [inline src/data/vocab_base.h:13]: You now have two identical typedefs.  Not good.

File: src/data/vocab_base.h
Comment (kpu) [inline src/data/vocab_base.h:43]: Don't put variable names here because clang will complain about unused function arguments.  It should just be `const string_view &, std::vector<string_view> &, ...)

File: src/data/vocab_base.h
Comment (kpu) [inline src/data/vocab_base.h:12]: Does this compile without sentencepiece?  It's a bit messy but probably ok.

File: src/data/vocab_base.h
Comment (jerinphilip) [reply src/data/vocab_base.h:43]: Note to self: Body doesn't use variables, which raises warning.

**Comment by jerinphilip** [comment]
Regression test using the test-app checked is stashed here at the moment, and works as intended.
https://github.com/jerinphilip/marian-regression-tests/commit/490b4c0b0f2edc02a1e0d146242fd68c422898e4

File: src/data/vocab.h
Comment (kpu) [inline src/data/vocab.h:55]: Nit: spaces around =

Comment (jerinphilip) [reply src/data/vocab.h]: https://github.com/browsermt/marian-dev/blob/2fbe6370b86157639c106f7c764a0f4017179beb/src/data/vocab.h#L53-L56

**Comment by jerinphilip** [comment]
`encodePreservingSource` -> `encodeWithByteRanges` (This can be applied on target vocabulary as well in theory; vocab is agnostic to source or target). `alignments` -> `byteRanges`. Name-changes will propagate into [bergamot-translator](https://github.com/browsermt/bergamot-translator/pull/8) soon, hopefully eliminating confusion.

Test-app has been modified accordingly, also regression-tests-repo. The test-script passes (for this feature) as indicated below. Is PR merge ready? 

```
jerin@gna:/fs/bil0/jerin/code/marian-regression-tests$ MARIAN=../marian-dev/build CUDA_VISIBLE_DEVICES=1 ./run_mrt.sh tests/sentencepiece/test_sentencepiece_encodeWithByteRanges.sh
[01/27/2021 20:54:39] Running on gna as process 8757
[01/27/2021 20:54:39] Using Marian binary: /mnt/bil0/jerin/code/marian-dev/build/marian
[01/27/2021 20:54:39] Version: v1.9.56 caaf365 2021-01-27 15:49:27 +0000
[01/27/2021 20:54:39] Build type: Release
[01/27/2021 20:54:39] Using compiler: /usr/bin/c++
[01/27/2021 20:54:39] Using MKL: COMPILE_CPU=on
[01/27/2021 20:54:39] Using CUDNN:
[01/27/2021 20:54:39] Using SentencePiece: USE_SENTENCEPIECE=ON
[01/27/2021 20:54:39] Using FBGEMM:
[01/27/2021 20:54:39] Unit tests: COMPILE_TESTS=ON
[01/27/2021 20:54:39] Using CUDA visible devices: 1
[01/27/2021 20:54:39] Using number of GPU devices: 1
[01/27/2021 20:54:39] Using time out: 5m
[01/27/2021 20:54:39] Checking directory: tests/sentencepiece
[01/27/2021 20:54:39] Running setup script
[01/27/2021 20:54:40] Running tests/sentencepiece/test_sentencepiece_encodeWithByteRanges.sh ... OK
[01/27/2021 20:54:40] Test took 00:00:0.569s
---------------------
Ran 1 tests in 00:00:0.944s, 1 passed, 0 skipped, 0 failed
```

browsermt/marian-dev with regression-tests are failing in several other tests:
* http://vali.inf.ed.ac.uk/jenkins/job/browsermt-marian-regression-tests/6/console
Requesting @kpu to take a look.

File: src/data/sentencepiece_vocab.cpp
Comment (kpu) [inline src/data/sentencepiece_vocab.cpp:228]: for (auto piece : spt.pieces()) {

**Review by kpu — APPROVED** [review]
Nits

**Comment by kpu** [comment]
@jerinphilip wanted to do a rebase before merge.

**Comment by abhi-agg** [comment]
@kpu @jerinphilip Could we merge this change to master if it is approved?

**Pull request merged**
